### PR TITLE
Enable int4 quantization for parakeet

### PR DESCRIPTION
**Pull request by larryliu0820**
## Summary
- Add int4/int8 quantization support for Parakeet TDT model export using torchao
- Extract quantization utilities to a separate module for reusability

### Quantization Support for Parakeet
Added support for quantizing encoder and decoder components with multiple configurations:
- **Linear layers**: `4w`, `8w`, `8da4w`, `8da8w` quantization configs
- **Embedding layers**: `4w`, `8w` quantization configs  
- **Packing formats**: `tile_packed_to_4d` for optimized inference on CUDA

#### New CLI Arguments
| Argument | Description |
|----------|-------------|
| `--qlinear_encoder` | Quantization config for encoder linear layers |
| `--qlinear_encoder_group_size` | Group size for encoder quantization (default: 32) |
| `--qlinear_encoder_packing_format` | Packing format for encoder |
| `--qlinear` | Quantization config for decoder linear layers |
| `--qlinear_group_size` | Group size for decoder quantization (default: 32) |
| `--qlinear_packing_format` | Packing format for decoder |
| `--qembedding` | Quantization config for embedding layer |
| `--qembedding_group_size` | Group size for embedding quantization |

### Code Organization
- Extracted `quantize_model_()` function to `examples/models/parakeet/quantize.py` (mirrors optimum-executorch naming)
- Model moved to CUDA after preprocessor export when `--backend cuda` is specified
- Example inputs created on correct device to match model parameters

### Int4 Linear with Tile Packing
```bash
python examples/models/parakeet/export_parakeet_tdt.py \
    --backend cuda \
    --dtype bf16 \
    --qlinear_encoder 4w \
    --qlinear_encoder_packing_format tile_packed_to_4d \
    --qlinear 4w \
    --qlinear_packing_format tile_packed_to_4d \
    --output-dir ./parakeet_int4
```

### Int4 Linear + Int8 Embedding
```bash
python examples/models/parakeet/export_parakeet_tdt.py \
    --backend cuda \
    --dtype bf16 \
    --qlinear_encoder 4w \
    --qlinear_encoder_packing_format tile_packed_to_4d \
    --qlinear 4w \
    --qlinear_packing_format tile_packed_to_4d \
    --qembedding 8w \
    --output-dir ./parakeet_int4_emb8
```

### Test Plan
- [x] Export with CUDA backend and int4 quantization completes successfully 
- [x] Model runs through encoder with storage_offset tensors 
- [x] Verify full transcription accuracy matches eager mode 
- [x] Verify model size reduction with quantization

### bfloat16 (baseline)
```
-rw-r--r--. 1 dev dev 1274592908 Jan 23 07:01 aoti_cuda_blob.ptd
-rw-r--r--. 1 dev dev    4497576 Jan 23 07:01 model.pte
```

### Encoder and decoder int4 group (32) wise quantized
```
-rw-r--r--. 1 dev dev 542162572 Jan 30 07:03 aoti_cuda_blob.ptd
-rw-r--r--. 1 dev dev   4125496 Jan 30 07:03 model.pte
```

### Int4 linear + Int8 embedding
```
-rw-r--r--.  1 dev dev 536943756 Feb  2 18:21 aoti_cuda_blob.ptd
-rw-r--r--.  1 dev dev   4134072 Feb  2 18:21 model.pte
```

### CODE REVIEW

File: examples/models/parakeet/quantize.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+                return m.weight.shape[1] % qlinear_group_size == 0
+            return False
+
+        print(
+            f"  Applying {qlinear_config} linear quantization "
+            f"(group_size={qlinear_group_size}, packing={qlinear_packing_format})..."
+        )
+        quantize_(module, config, filter_fn=linear_filter)
+
+    unwrap_tensor_subclass(module)
```
Comment (mergennachin) [inline examples/models/parakeet/quantize.py]: In optimum-executorch, it is conditionally called only for PyTorch < 2.10.0. We can skip here?

File: examples/models/parakeet/quantize.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
+"""Quantization utilities for Parakeet model export."""
+
+from typing import Optional
+
+import torch
+
+
+def quantize(  # noqa: C901
```
Comment (mergennachin) [inline examples/models/parakeet/quantize.py]: Can you rename to quantize_model_ to mirror optimum-executorch

File: examples/models/parakeet/quantize.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
+"""Quantization utilities for Parakeet model export."""
```
Comment (mergennachin) [inline examples/models/parakeet/quantize.py:1]: This is basically optimum-executorch, right? There's nothing really specific to Parakeet

Jack/Kimish was planning to bring in all the optimum logic back into executorch, and have optimum-executorch to be a very thin wrapper.

Not in this PR, but in the future, I can image we move this `extension/export_util/quantize.py` or `extension/quantization/torchao_utils.py`

File: examples/models/parakeet/quantize.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+            )
+        else:
+            raise ValueError(f"Unsupported qlinear_config: {qlinear_config}")
+
+        # Filter: only quantize Linear layers with compatible dimensions
+        def linear_filter(m, fqn):
+            if isinstance(m, torch.nn.Linear):
+                if qlinear_group_size == 0:
+                    return True
+                return m.weight.shape[1] % qlinear_group_size == 0
```
Comment (mergennachin) [inline examples/models/parakeet/quantize.py]: Log if it isn't compatible

File: .github/workflows/cuda.yml [hunk 5]
```diff
@@ -150,11 +150,6 @@ jobs:
               repo: "google"
               name: "gemma-3-4b-it"
             quant: "quantized-int4-weight-only"
-          # Parakeet only supports non-quantized
-          - model:
-              repo: "nvidia"
-              name: "parakeet-tdt"
-            quant: "quantized-int4-tile-packed"
```
Comment (mergennachin) [inline .github/workflows/cuda.yml:157]: why not delete the one below? line 158-161


also, we still have exclusion for  test-model-cuda-e2e around line ~210

https://github.com/pytorch/executorch/blob/9d257c81a5850d6b285bfb62805d591e5efc1143/.github/workflows/cuda.yml#L222-L230

File: examples/models/parakeet/README.md [hunk 6]
```diff
@@ -31,6 +31,48 @@ python export_parakeet_tdt.py --audio /path/to/audio.wav
...
+|--------|-------------|
+| `4w` | 4-bit weight only quantization |
+| `8w` | 8-bit weight only quantization |
+| `8da4w` | 8-bit dynamic activation, 4-bit weight |
+| `8da8w` | 8-bit dynamic activation, 8-bit weight |
+
+#### Example: 4-bit Weight Quantization with Tile Packing
+
+```bash
+python export_parakeet_tdt.py \
```
Comment (mergennachin) [inline examples/models/parakeet/README.md:63]: does the embedding quantization work? we can start with 8w and per-axis

**Review by mergennachin — CHANGES_REQUESTED** [review]

**Comment by mergennachin** [comment]
Also please report size reductions 

- encoder quantized (4w, groupwise)
- encoder and decoder quantized (4w, groupwise)
- encoder, decoder (4w, groupwise) and embedding quantized (8w, per-channel)

File: examples/models/parakeet/quantize.py (hunk 3 above)
Comment (larryliu0820) [reply examples/models/parakeet/quantize.py:1]: Yeah this should be reusable.

File: examples/models/parakeet/README.md [hunk 7]
```diff
@@ -31,6 +31,48 @@ python export_parakeet_tdt.py --audio /path/to/audio.wav
...
+    --backend xnnpack \
```
Comment (mergennachin) [inline examples/models/parakeet/README.md]: In the example, use CUDA

4w tile_packed_to_4d is not compatible with xnnpack

File: examples/models/parakeet/README.md [hunk 8]
```diff
@@ -31,6 +31,48 @@ python export_parakeet_tdt.py --audio /path/to/audio.wav
...
+    --qlinear_encoder 4w \
+    --qlinear_encoder_group_size 32 \
+    --qlinear_encoder_packing_format tile_packed_to_4d \
+    --qlinear 4w \
+    --qlinear_group_size 32 \
+    --qlinear_packing_format tile_packed_to_4d \
+    --output-dir ./parakeet_quantized
+```
+
+**Note:** The `tile_packed_to_4d` packing format is optimized for certain backends like XNNPACK, CUDA, and Metal.
```
Comment (mergennachin) [inline examples/models/parakeet/README.md]: Is tile_packed_to_4d even compatible with xnnpack?

**Review by mergennachin — APPROVED** [review]

**Review by mergennachin — APPROVED** [review]
